File: packages/asdl-objectives/src/asdl_objectives/exec/read_objective.py
"""``objective exec read-objective`` filesystem reader."""

from __future__ import annotations

from pathlib import Path
from typing import Annotated, Literal

import click

from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from asdl_objectives.objective_storage import (
    FilesystemObjectiveStorage,
    ObjectiveFiles,
    ObjectiveUpdateFile,
    active_record_relative_path,
    active_root_relative_path,
    empty_objective_files,
    is_valid_objective_slug,
    render_file_presence,
)

ReadObjectiveStatus = Literal["ok", "missing_slug", "invalid_slug", "not_found"]


class ReadObjectiveRequest(ClinkrModel):
    slug: Annotated[
        str | None,
        click.Argument(["slug"], type=click.STRING, required=False, default=None),
    ] = None


class ReadObjectiveResult(ClinkrModel):
    status: ReadObjectiveStatus
    error: str | None
    root_path: str
    root_exists: bool
    slug: str | None
    path: str | None
    exists: bool
    closed: bool
    files: ObjectiveFiles
    updates: tuple[ObjectiveUpdateFile, ...]
    update_count: int


def render_read_objective(result: ReadObjectiveResult) -> None:
    if result.slug is None or result.path is None:
        click.echo("No Objective record selected.")
        return

    record_path = Path.cwd() / result.path
    root_state = "present" if result.root_exists else "missing"
    state = "closed" if result.closed else "open"

    click.echo(f"# Objective `{result.slug}`")
    click.echo()
    click.echo(f"Root: `{result.root_path}` ({root_state})")
    click.echo(f"Path: `{result.path}`")
    click.echo(f"State: {state}")
    click.echo(f"Files: {render_file_presence(result.files)}")
    click.echo(f"Updates: {result.update_count}")
    click.echo()

    storage = FilesystemObjectiveStorage(Path.cwd())
    _render_markdown_file(storage, record_path / "objective.md", "objective.md")
    _render_markdown_file(storage, record_path / "roadmap.md", "roadmap.md")
    _render_updates(storage, result, record_path)


@clinkr_operation(
    name="read-objective",
    help="Read one Objective record by explicit slug as filesystem facts or raw Markdown.",
    human_renderer=render_read_objective,
)
def run_read_objective(
    ctx: click.Context,
    request: ReadObjectiveRequest,
) -> ClinkrExit[ReadObjectiveResult]:
    del ctx
    storage = FilesystemObjectiveStorage(Path.cwd())
    root = active_root_relative_path()
    root_exists = storage.active_root_path().exists()

    if request.slug is None:
        raise ClinkrExit.negative(
            _empty_result(
                status="missing_slug",
                error="missing_slug",
                slug=None,
                path=None,
                root_exists=root_exists,
            ),
            message="Missing Objective slug. Pass an explicit slug.",
        )

    if not is_valid_objective_slug(request.slug):
        raise ClinkrExit.negative(
            _empty_result(
                status="invalid_slug",
                error="invalid_slug",
                slug=None,
                path=None,
                root_exists=root_exists,
            ),
            message=f"Invalid Objective slug {request.slug!r}. Pass a single slug, not a path.",
        )

    relative_path = active_record_relative_path(request.slug)
    absolute_path = storage.active_record_path(request.slug)
    if not absolute_path.is_dir():
        raise ClinkrExit.negative(
            _empty_result(
                status="not_found",
                error="not_found",
                slug=request.slug,
                path=relative_path.as_posix(),
                root_exists=root_exists,
            ),
            message=f"No Objective record found for slug {request.slug!r}.",
        )

    files = storage.file_presence(absolute_path)
    updates = storage.list_update_files(absolute_path)
    return ClinkrExit.ok(
        ReadObjectiveResult(
            status="ok",
            error=None,
            root_path=root.as_posix(),
            root_exists=root_exists,
            slug=request.slug,
            path=relative_path.as_posix(),
            exists=True,
            closed=files.closed_md,
            files=files,
            updates=updates,
            update_count=len(updates),
        )
    )


def _empty_result(
    *,
    status: ReadObjectiveStatus,
    error: str,
    slug: str | None,
    path: str | None,
    root_exists: bool,
) -> ReadObjectiveResult:
    return ReadObjectiveResult(
        status=status,
        error=error,
        root_path=active_root_relative_path().as_posix(),
        root_exists=root_exists,
        slug=slug,
        path=path,
        exists=False,
        closed=False,
        files=empty_objective_files(),
        updates=(),
        update_count=0,
    )


def _render_markdown_file(
    storage: FilesystemObjectiveStorage,
    path: Path,
    display_path: str,
) -> None:
    click.echo(f"## {display_path}")
    click.echo()
    content = storage.read_markdown_file(path)
    if content is None:
        click.echo(f"_Missing `{display_path}`._")
        click.echo()
        return

    click.echo(content, nl=False)
    if not content.endswith("\n"):
        click.echo()
    click.echo()


def _render_updates(
    storage: FilesystemObjectiveStorage,
    result: ReadObjectiveResult,
    record_path: Path,
) -> None:
    if not result.files.updates_dir:
        click.echo("## updates/")
        click.echo()
        click.echo("_Missing `updates/` directory._")
        click.echo()
        return

    if not result.updates:
        click.echo("## updates/")
        click.echo()
        click.echo("_No direct update Markdown files found._")
        click.echo()
        return

    for update in result.updates:
        _render_markdown_file(
            storage,
            record_path / "updates" / update.name,
            f"updates/{update.name}",
        )

File: ts/packages/objective/src/operations/read-objective.ts
import { failure, negative, ok, type ClinkrExit } from "@asdl/clinkr";
import { z } from "zod";

import type { ObjectiveCliContext } from "../context.ts";
import { pythonStringRepr, removeOneTrailingNewline } from "./format.ts";
import { handleObjectiveSlugValidationErrors } from "./slug-validation-errors.ts";
import {
	activeRecordRelativePath,
	activeRootRelativePath,
	emptyObjectiveFiles,
	isValidObjectiveSlug,
	renderFilePresence,
	type ObjectiveFiles,
	type ObjectiveMarkdownReadResult,
	type ObjectiveStorage,
	type ObjectiveUpdateFile,
} from "../storage.ts";

export const readObjectiveRequestSchema = z.object({
	slug: z.string().optional().describe("Objective slug to read."),
});

export const objectiveFilesSchema = z.object({
	objectiveMd: z.boolean(),
	roadmapMd: z.boolean(),
	updatesDir: z.boolean(),
	closedMd: z.boolean(),
});

export const objectiveUpdateFileSchema = z.object({
	name: z.string(),
	path: z.string(),
});

export const objectiveMarkdownReadResultSchema = z.discriminatedUnion("type", [
	z.object({ type: z.literal("missing") }),
	z.object({ type: z.literal("ok"), content: z.string() }),
	z.object({ type: z.literal("unreadable"), message: z.string() }),
]);

export const readObjectiveBaseResultSchema = z.object({
	error: z.string().nullable(),
	rootPath: z.string(),
	rootExists: z.boolean(),
	slug: z.string().nullable(),
	path: z.string().nullable(),
	exists: z.boolean(),
	closed: z.boolean(),
	files: objectiveFilesSchema,
	updates: z.array(objectiveUpdateFileSchema),
	updateCount: z.number().int(),
});

export const readObjectiveNonOkResultSchema = z.discriminatedUnion("status", [
	readObjectiveBaseResultSchema.extend({ status: z.literal("missing_slug") }),
	readObjectiveBaseResultSchema.extend({ status: z.literal("invalid_slug") }),
	readObjectiveBaseResultSchema.extend({ status: z.literal("not_found") }),
]);

export const readObjectiveOkResultSchema = readObjectiveBaseResultSchema.extend({
	status: z.literal("ok"),
	markdownFiles: z.object({
		objectiveMd: objectiveMarkdownReadResultSchema,
		roadmapMd: objectiveMarkdownReadResultSchema,
		updates: z.array(
			z.object({
				update: objectiveUpdateFileSchema,
				content: objectiveMarkdownReadResultSchema,
			}),
		),
	}),
});

export const readObjectiveResultSchema = z.discriminatedUnion("status", [
	readObjectiveOkResultSchema,
	readObjectiveBaseResultSchema.extend({ status: z.literal("missing_slug") }),
	readObjectiveBaseResultSchema.extend({ status: z.literal("invalid_slug") }),
	readObjectiveBaseResultSchema.extend({ status: z.literal("not_found") }),
]);

export type ReadObjectiveRequest = z.infer<typeof readObjectiveRequestSchema>;
export type ReadObjectiveStatus = "ok" | "missing_slug" | "invalid_slug" | "not_found";
export type ReadObjectiveResult = z.infer<typeof readObjectiveResultSchema>;

interface ReadObjectiveMarkdownFiles {
	objectiveMd: ObjectiveMarkdownReadResult;
	roadmapMd: ObjectiveMarkdownReadResult;
	updates: readonly { update: ObjectiveUpdateFile; content: ObjectiveMarkdownReadResult }[];
}

export interface ReadObjectiveOkResult extends ReadObjectiveBaseResult {
	status: "ok";
	error: null;
	slug: string;
	path: string;
	exists: true;
	markdownFiles: ReadObjectiveMarkdownFiles;
}

interface ReadObjectiveBaseResult {
	status: ReadObjectiveStatus;
	error: string | null;
	rootPath: string;
	rootExists: boolean;
	slug: string | null;
	path: string | null;
	exists: boolean;
	closed: boolean;
	files: ObjectiveFiles;
	updates: readonly ObjectiveUpdateFile[];
	updateCount: number;
}

export async function runReadObjective(
	ctx: ObjectiveCliContext,
	request: ReadObjectiveRequest,
): Promise<ClinkrExit<ReadObjectiveResult>> {
	const result = await readObjective(ctx.storage, request.slug);
	if (result.type === "storage-error") return failure(result.error.code, result.error.message);
	const slugValidationError = handleObjectiveSlugValidationErrors(result.value, request.slug);
	if (slugValidationError !== null) return slugValidationError;
	if (result.value.status === "not_found") {
		return negative(
			`No Objective record found for slug ${pythonStringRepr(result.value.slug ?? "")}.`,
			result.value,
		);
	}
	return ok(result.value);
}

export function renderReadObjective(result: ReadObjectiveResult): string {
	if (result.status !== "ok") return "No Objective record selected.";

	const rootState = result.rootExists ? "present" : "missing";
	const state = result.closed ? "closed" : "open";
	const parts = [
		`# Objective \`${result.slug}\`\n\n`,
		`Root: \`${result.rootPath}\` (${rootState})\n`,
		`Path: \`${result.path}\`\n`,
		`State: ${state}\n`,
		`Files: ${renderFilePresence(result.files)}\n`,
		`Updates: ${result.updateCount}\n\n`,
	];

	appendMarkdownFile(parts, "objective.md", result.markdownFiles.objectiveMd);
	appendMarkdownFile(parts, "roadmap.md", result.markdownFiles.roadmapMd);
	if (!result.files.updatesDir) {
		parts.push("## updates/\n\n_Missing `updates/` directory._\n\n");
	} else if (result.markdownFiles.updates.length === 0) {
		parts.push("## updates/\n\n_No direct update Markdown files found._\n\n");
	} else {
		for (const update of result.markdownFiles.updates) {
			appendMarkdownFile(parts, `updates/${update.update.name}`, update.content);
		}
	}
	return removeOneTrailingNewline(parts.join(""));
}

async function readObjective(
	storage: ObjectiveStorage,
	slug: string | undefined,
): Promise<
	| { type: "ok"; value: ReadObjectiveResult }
	| { type: "storage-error"; error: { code: string; message: string } }
> {
	const root = activeRootRelativePath();
	const rootPresence = await storage.activeRootExists();
	if (!rootPresence.ok) return { type: "storage-error", error: rootPresence.error };

	if (slug === undefined) {
		return {
			type: "ok",
			value: emptyResult({
				status: "missing_slug",
				error: "missing_slug",
				root,
				slug: null,
				path: null,
				hasRoot: rootPresence.value,
			}),
		};
	}

	if (!isValidObjectiveSlug(slug)) {
		return {
			type: "ok",
			value: emptyResult({
				status: "invalid_slug",
				error: "invalid_slug",
				root,
				slug: null,
				path: null,
				hasRoot: rootPresence.value,
			}),
		};
	}

	const relativePath = activeRecordRelativePath(slug);
	const exists = await storage.activeRecordExists(slug);
	if (!exists.ok) return { type: "storage-error", error: exists.error };
	if (!exists.value) {
		return {
			type: "ok",
			value: emptyResult({
				status: "not_found",
				error: "not_found",
				root,
				slug,
				path: relativePath,
				hasRoot: rootPresence.value,
			}),
		};
	}

	const files = await storage.filePresence(relativePath);
	if (!files.ok) return { type: "storage-error", error: files.error };
	const updates = await storage.listUpdateFiles(relativePath);
	if (!updates.ok) return { type: "storage-error", error: updates.error };
	const facts = {
		status: "ok" as const,
		error: null,
		rootPath: root,
		rootExists: rootPresence.value,
		slug,
		path: relativePath,
		exists: true,
		closed: files.value.closedMd,
		files: files.value,
		updates: [...updates.value],
		updateCount: updates.value.length,
	};
	return {
		type: "ok",
		value: {
			...facts,
			markdownFiles: {
				objectiveMd: await storage.readMarkdownFile(`${relativePath}/objective.md`),
				roadmapMd: await storage.readMarkdownFile(`${relativePath}/roadmap.md`),
				updates: await Promise.all(
					updates.value.map(async (update) => ({
						update,
						content: await storage.readMarkdownFile(`${relativePath}/updates/${update.name}`),
					})),
				),
			},
		},
	};
}

function emptyResult(options: {
	status: Exclude<ReadObjectiveStatus, "ok">;
	error: string;
	root: string;
	slug: string | null;
	path: string | null;
	hasRoot: boolean;
}): ReadObjectiveResult {
	return {
		status: options.status,
		error: options.error,
		rootPath: options.root,
		rootExists: options.hasRoot,
		slug: options.slug,
		path: options.path,
		exists: false,
		closed: false,
		files: emptyObjectiveFiles(),
		updates: [],
		updateCount: 0,
	};
}

function appendMarkdownFile(
	parts: string[],
	displayPath: string,
	read: ObjectiveMarkdownReadResult,
): void {
	parts.push(`## ${displayPath}\n\n`);
	if (read.type === "missing") {
		parts.push(`_Missing \`${displayPath}\`._\n\n`);
		return;
	}
	if (read.type === "unreadable") {
		parts.push(`_Unable to read \`${displayPath}\`: ${read.message}_\n\n`);
		return;
	}
	parts.push(read.content);
	if (!read.content.endsWith("\n")) parts.push("\n");
	parts.push("\n");
}
